### PR TITLE
fixtodo:rsDeepCopy only when sizeNeedsUpdate or annotationsNeedUpdate

### PR DESCRIPTION
**Pull request by hanxiaoshuai**
**What this PR does / why we need it**:
```
// TODO: Do not mutate the replica set here, instead simply compare the annotation and if they mismatch
// call SetReplicasAnnotations inside the following if clause. Then we can also move the deep-copy from
// above inside the if too.
```
fixtodo:rsDeepCopy only when sizeNeedsUpdate or annotationsNeedUpdate
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/controller/deployment/util/deployment_util.go [hunk 1]
```diff
@@ -400,6 +400,22 @@ func SetReplicasAnnotations(rs *extensions.ReplicaSet, desiredReplicas, maxRepli
 	return updated
 }
 
+// AnnotationsNeedUpdate return true if ReplicasAnnotations need to be updated
+func AnnotationsNeedUpdate(rs *extensions.ReplicaSet, desiredReplicas, maxReplicas int32) bool {
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: ReplicasAnnotationsNeedUpdate

File: pkg/controller/deployment/sync.go [hunk 2]
```diff
@@ -402,18 +402,19 @@ func (dc *DeploymentController) scaleReplicaSetAndRecordEvent(rs *extensions.Rep
...
+	sizeNeedsUpdate := *(rs.Spec.Replicas) != newScale
+
+	annotationsNeedUpdate := deploymentutil.AnnotationsNeedUpdate(rs, *(deployment.Spec.Replicas), *(deployment.Spec.Replicas)+deploymentutil.MaxSurge(*deployment))
 
 	scaled := false
 	var err error
 	if sizeNeedsUpdate || annotationsNeedUpdate {
+		rsCopy := rs.DeepCopy()
 		*(rsCopy.Spec.Replicas) = newScale
+		if annotationsNeedUpdate {
```
Comment (kargakis) [inline pkg/controller/deployment/sync.go]: Do this unconditionally similar to how the scale size is set unconditionally.

File: pkg/controller/deployment/util/deployment_util.go (hunk 1 above)
Comment (hanxiaoshuai) [reply pkg/controller/deployment/util/deployment_util.go]: Done

File: pkg/controller/deployment/sync.go (hunk 2 above)
Comment (hanxiaoshuai) [reply pkg/controller/deployment/sync.go]: Done,Thanks for your review.

**Pull request merged**